crates/dbsp/src/trace/ord/fallback/key_batch.rs

```diff
@@ -426,41 +436,62 @@ where
     }
 }
 
-impl<K, T, R> SizeOf for FallbackKeyMerger<K, T, R>
+/// A builder for creating layers from unsorted update tuples.
+#[derive(SizeOf)]
+pub struct FallbackKeyBuilder<K, T, R>
 where
     K: DataTrait + ?Sized,
     T: Timestamp,
     R: WeightTrait + ?Sized,
 {
-    fn size_of_children(&self, context: &mut size_of::Context) {
-        match &self.inner {
-            MergerInner::AllFile(file) => file.size_of_children(context),
-            MergerInner::AllVec(vec) => vec.size_of_children(context),
-            MergerInner::ToFile(merger) => merger.size_of_children(context),
-            MergerInner::ToVec(merger) => merger.size_of_children(context),
-        }
-    }
+    #[size_of(skip)]
+    factories: FallbackKeyBatchFactories<K, T, R>,
+    inner: BuilderInner<K, T, R>,
 }
 
-/// A builder for creating layers from unsorted update tuples.
-pub struct FallbackKeyBuilder<K, T, R>
+#[derive(SizeOf)]
+enum BuilderInner<K, T, R>
 where
     K: DataTrait + ?Sized,
     T: Timestamp,
     R: WeightTrait + ?Sized,
 {
-    factories: FallbackKeyBatchFactories<K, T, R>,
-    inner: BuilderInner<K, T, R>,
+    // In-memory.
+    Vec(VecKeyBuilder<K, T, R>),
+
+    // On-storage.
+    File(FileKeyBuilder<K, T, R>),
+
+    /// In-memory as long as we don't exceed a maximum threshold size.
+    Threshold {
+        vec: VecKeyBuilder<K, T, R>,
+
+        /// Bytes left to add until the threshold is exceeded.
+        remaining: usize,
+    },
 }
 
-enum BuilderInner<K, T, R>
+impl<K, T, R> FallbackKeyBuilder<K, T, R>
 where
+    Self: SizeOf,
     K: DataTrait + ?Sized,
     T: Timestamp,
     R: WeightTrait + ?Sized,
 {
-    File(FileKeyBuilder<K, T, R>),
-    Vec(VecKeyBuilder<K, T, R>),
+    /// We ran out of the bytes threshold for `BuilderInner::Threshold`. Spill
+    /// to storage.
+    fn over_threshold(&mut self) {
+        let new_inner =
+            BuilderInner::File(FileKeyBuilder::timed_with_capacity(&self.factories.file, 0));
+        let BuilderInner::Threshold { vec, .. } = replace(&mut self.inner, new_inner) else {
+            unreachable!()
```

Comment:
might be good to add a comment reasoning why this should never be true in case it ever happens to go there in the future

Reply:
I came up with a way that doesn't require this.